python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -0,0 +1,135 @@
+import pathlib
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Union
+
+import imageio as iio
+import numpy as np
+from ray.data.block import Block
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.data.datasource.datasource import ReadTask
+from ray.data.datasource.file_based_datasource import _resolve_paths_and_filesystem
+from ray.data.datasource.file_meta_provider import (
+    BaseFileMetadataProvider,
+    DefaultFileMetadataProvider,
+    FastFileMetadataProvider,
+)
+from ray.data.datasource.partitioning import PathPartitionFilter
+
+if TYPE_CHECKING:
+    import pyarrow
+
+IMAGE_EXTENSIONS = [".png", ".jpg", ".jpeg", ".tiff", ".bmp", ".gif"]
+
+
+class ImageFolderDatasource(BinaryDatasource):
+    """A datasource that lets you read datasets like `ImageNet <https://www.image-net.org/>`_.
+
+    This datasource works with any dataset where images are arranged in this way:
+
+    .. code-block::
+
+        root/dog/xxx.png
+        root/dog/xxy.png
+        root/dog/[...]/xxz.png
+
+        root/cat/123.png
+        root/cat/nsdf3.png
+        root/cat/[...]/asd932_.png
+
```

Comment:
nit: Call out that `dog` and `cat` are labels?

Reply:
Do you feel like it's not obvious that `dog` and `cat` are the labels? I feel like based on convention it should be clear, but maybe I'm suffering from the curse of knowledge.